File: src/com/minhle/practiceleetcode/TrappingRainWater.java
package com.minhle.practiceleetcode;

import org.junit.jupiter.api.Test;

public class TrappingRainWater {
    int[] maxLeft;
    int[] maxRight;
    int[] minLeftRight;

    private void calculateHelperArrays(int[] input) {
        maxLeft = new int[input.length];
        maxRight = new int[input.length];
        minLeftRight = new int[input.length];
        findMaxFromLeft(input);
        findMaxFromRight(input);
        findMinLeftRight(input);
    }
    private void findMaxFromLeft(int[] input) {
        for (int i = 1; i < maxLeft.length; i++) {
            maxLeft[i] = Math.max(maxLeft[i - 1], input[i - 1]);
        }
    }
    private void findMaxFromRight(int[] input) {
        for (int i = maxRight.length - 2; i >= 0; i--) {
            maxRight[i] = Math.max(maxRight[i + 1], input[i + 1]);
        }
    }

    private void findMinLeftRight(int[] input) {
        for (int i = 0; i < minLeftRight.length; i++) {
            minLeftRight[i] = Math.min(maxLeft[i], maxRight[i]);
        }
    }

    private int calculateTotalTrappedWater(int[] input) {
        calculateHelperArrays(input);
        int[] totalTrappedWater = new int[input.length];
        for (int i = 0; i < totalTrappedWater.length; i++) {
            int water = minLeftRight[i] - input[i];
            totalTrappedWater[i] = Math.max(water, 0);
        }
        int sum = 0;
        for (int water : totalTrappedWater) {
            sum += water;
        }
        return sum;
    }

    public int trap(int[] height) {
        return calculateTotalTrappedWater(height);
    }

    @Test
    void test1() {
        int[] input = {0,1,0,2,1,0,1,3,2,1,2,1};
        calculateTotalTrappedWater(input);
    }
}

File: src/com/minhle/practiceleetcode/TrappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        auto max_left = find_max_left(height);
        auto max_right = find_max_right(height);
        int res = 0;
        for (int i = 0; i < height.size(); ++i)
        {
            res += max(0, min(max_left[i], max_right[i]) - height[i]);
        }
        return res;
    }
private:
    vector<int> find_max_right(const vector<int> &heights)
    {
        vector<int> max_right(heights.size(), 0);
        for (int i = heights.size() - 2; i >= 0; --i)
        {
            max_right[i] = max(max_right[i + 1], heights[i + 1]);
        }
        return max_right;
    }

    vector<int> find_max_left(const vector<int> &heights)
    {
        vector<int> max_left(heights.size(), 0);
        for (int i = 1; i < heights.size(); ++i)
        {
            max_left[i] = max(max_left[i - 1], heights[i - 1]);
        }
        return max_left;
    }
};
